### ISSUE TITLE
datastore: Saving string of hash causes Half-closed rpc error

### ISSUE THREAD

**Issue by fanktom** [GoogleCloudPlatform/google-cloud-go#632]
Hey there,

I recently ran into a problem where I was trying to save a string of a hash as a property.
See for example the following code snippet:

```
package main

import (
	"context"
	"crypto/sha256"
	"io"
	"log"
	"os"

	"cloud.google.com/go/datastore"
)

type Entity struct {
	Property string
}

func main() {
	client, err := datastore.NewClient(context.Background(), "foo")
	if err != nil {
		log.Fatal(err)
	}
	k := datastore.NameKey("baz", "foo", nil)

	h := sha256.New()
	io.WriteString(h, "bar")
	e := Entity{Property: string(h.Sum(nil))}

	_, err = client.Put(context.Background(), k, &e)
	if err != nil {
		log.Fatal(err)
	}
}
```

Running this results in an
```
rpc error: code = Internal desc = Half-closed without a request
```

If you change the string conversion to e.g. base 16, with lower-case letters for a-f like this:
```
fmt.Sprintf("%x", h.Sum(nil))
```
it works.

Now I am not sure if this is desired behaviour. One could probably argue that the string contains some weird sequences it should not, which then terminate the connection. But then again I have a feeling that it should not be possible to kill an underlying HTTP connection by sending a perfectly valid string.

Further, this could be a grpc issue unrelated to datastore itself, which I can confirm if this is undersired behaviour.

What do you think?

**Comment by rogerhub** [comment]
It might be that strings are meant to contain valid UTF-8 (and perhaps invalid sequences are replaced with U+FFFD). Is there any problem with using []byte instead? If you're storing the raw bytes of a hash, it should probably be in a []byte.

**Comment by fanktom** [comment]
Thanks @rogerhub, yes, thats exactly what I am doing now. 

Still I find it a bit unexpected that the string is not masked or encoded during the conversion. I expected that the type safety of go's strings would ensure I can not disconnect an underlying HTTP connection but maybe that expectation is just wrong 😀

TLDR: Use a []byte type and don't convert to string.

**Comment by jba** [comment]
https://code-review.googlesource.com/c/13510 fixes this by disallowing strings that are not valid UTF-8.
